src/Features/CSharp/Portable/CodeRefactorings/UseExplicitOrImplicitType/AbstractUseTypeCodeRefactoringProvider.cs

```diff
@@ -85,11 +83,45 @@ public override async Task ComputeRefactoringsAsync(CodeRefactoringContext conte
                     c => UpdateDocumentAsync(document, declaredType, c)));
         }
 
-        private static SyntaxNode GetDeclaration(SyntaxNode root, TextSpan textSpan)
+        private static async Task<SyntaxNode> GetDeclarationAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
         {
-            var token = root.FindToken(textSpan.Start);
-            return token.Parent?.FirstAncestorOrSelf<SyntaxNode>(
-                a => a.IsKind(SyntaxKind.DeclarationExpression, SyntaxKind.VariableDeclaration, SyntaxKind.ForEachStatement));
+            // We want to provide refactoring for changing the Type of newly introduced variables in following cases:
+            // - DeclarationExpressionSyntax: `"42".TryParseInt32(out var number)`
+            // - VariableDeclarationSyntax: General field / variable declaration statement `var number = 42`
+            // - ForEachStatementSyntax: The variable that gets introduced by foreach `foreach(var number in numbers)`
+            //
+            // In addition to providing the refactoring when the whole node (i.e. the node that introduces the new variable) in question is selected 
+            // we also want to enable it when only the type node is selected because this refactoring changes the type. We still have to make sure 
+            // we're only working on TypeNodes for in above-mentioned situations.
+
+            var refactoringHelperService = document.GetLanguageService<IRefactoringHelpersService>();
+
+            var declNode = await refactoringHelperService.TryGetSelectedNodeAsync<DeclarationExpressionSyntax>(document, textSpan, cancellationToken).ConfigureAwait(false);
+            if (declNode != null)
+            {
+                return declNode;
+            }
+
+            var variableNode = await refactoringHelperService.TryGetSelectedNodeAsync<VariableDeclarationSyntax>(document, textSpan, cancellationToken).ConfigureAwait(false);
+            if (variableNode != null)
+            {
+                return variableNode;
+            }
+
+            var foreachStatement = await refactoringHelperService.TryGetSelectedNodeAsync<ForEachStatementSyntax>(document, textSpan, cancellationToken).ConfigureAwait(false);
+            if (foreachStatement != null)
+            {
+                return foreachStatement;
+            }
+
+            var typeNode = await refactoringHelperService.TryGetSelectedNodeAsync<TypeSyntax>(document, textSpan, cancellationToken).ConfigureAwait(false);
+            var typeNodeParent = typeNode?.Parent;
+            if (typeNodeParent != null && typeNodeParent.IsKind(SyntaxKind.DeclarationExpression, SyntaxKind.VariableDeclaration, SyntaxKind.ForEachStatement))
```

Comment:
don't need `!= null` with IsKind.  But i can see the desire to keep it.

Reply:
Oh, because it's an extension method and those don't nullcheck on invocation. I'll keep it however, seems more consistent with good practices.